src/model/gateway.rs

```diff
@@ -151,10 +152,9 @@ impl Activity {
     ///     Ok(())
     /// }
     /// ```
-    pub fn streaming<N, U>(name: N, url: U) -> Activity
+    pub fn streaming<N>(name: N, url: Url) -> Activity
```

Comment:
I find it wrong to force a dependency of Serenity upon users just so they can use this function (the `url` crate is not re-exported from Serenity, thus users will have to define `url` as a dependency in their bots). Could it be changed to a generic that tries to convert to a `Url`? The `Url` type has a `TryFrom<&str>` implementation, which we could use to abstract the construction of the `Url` from users and prevent breaking code as much as possible.

```rust
pub fn streaming<N, U>(name: N, url: U) -> Activity 
where
    N: ToString,
    U: TryInto<Url>,
{
    let url = url.try_into().expect("failed to parse url");
    // ....
}

Reply:
I don't see the harm in users needing to `use url;`. But I'm not opposed to ur suggestions either.

1. Should url be re-exported. If so, where?
2. `TryFrom` it is

Reply:
There is no harm per se, it just makes the library harder to set up, since you'll need to define `url` as a dependency besides `serenity` (see: [tokio](https://github.com/serenity-rs/serenity/blob/current/examples/e01_basic_ping_bot/Cargo.toml#L8-L9)). It also makes it likely for the `url` of Serenity and the `url` of the user to get out-of-date (for example, we update `url` to 3.0, but the user still depends on 2.2.2, which rustc will classify as separate crates, resulting in types that are incompatible with one another and thus compilation errors).

> Should url be re-exported. If so, where?

At the crate root, [just as it is done for other crates](https://github.com/serenity-rs/serenity/blob/current/src/lib.rs#L131-L137).

Reply:
I couldn't figure out what bound to use that would allow `Url::try_from` generically (in case they added impls for more than `&str`). I also don't understand Rust's Stringy conversions and types, but that's beyond me and this PR. I switched it back to `U: ToString` and internally use `Url::parse`. The signature is unchanged

Reply:
> I couldn't figure out what bound to use that would allow `Url::try_from` generically

That would be the `TryInto` trait I demonstrated in my example snippet. [`TryInto` has a generic implementation for every type that implements `TryFrom`](https://doc.rust-lang.org/nightly/src/core/convert/mod.rs.html#567-576), just like [the regular `Into` and `From` traits](https://doc.rust-lang.org/nightly/src/core/convert/mod.rs.html#533-540). If a type `Foo`, for example, implements `TryFrom<i32>`, then we can depend on this implementation indirectly with `TryInto<Foo>`. This will not only pick up the `TryFrom<i32>` implementation, but also any future `TryFrom` implementations on `Foo`.

Anyway, since we no longer use `Url` in the public API, and the API has not changed, you can retarget your PR to the `current` branch.

Reply:
I *think* then this is what I'm going for. This is back to a breaking change

```rust
use url::Url;

use std::convert::TryFrom;
use std::fmt::Debug;

pub fn streaming<N, U>(name: N, url: U) -> Activity
    where
        N: ToString,
        Url: TryFrom<U>,
        <Url as TryFrom<U>>::Error: Debug,

```